### ISSUE TITLE
What is the advantage of `sparse_stoch_mat` over a unmodified `csr_matrix` (or `csr_array`) representation?

### ISSUE THREAD

**Issue by j-i-l** [alexbovet/flow_stability#40]
With #35 I was looking into the [SparseStochMat module](https://github.com/alexbovet/flow_stability/blob/8648a8bd715e3b086ed13cb7453d41754c176398/src/flowstab/SparseStochMat.py) and was wondering what was the motivation for the [`sparse_sotch_mat`](https://github.com/alexbovet/flow_stability/blob/main/src/flowstab/SparseStochMat.py#L92) class.

I see that the class allows to reduce the footprint of a csr matrix as each row/column with a value only on the diagonal is represented implicitly by the `nz_rowcols` (absence of the index there) and `size` attributes.
On the other hand matrix operations like [`__rmatmul__`](https://github.com/alexbovet/flow_stability/blob/8648a8bd715e3b086ed13cb7453d41754c176398/src/flowstab/SparseStochMat.py#L667-LL675) need to be implemented and  require to rebuild the full csr matrix (which is some computational overhead), also undoing the reduced footprint (at least temporarily).

I'm asking since a row/col with all zeros except on the diagonal simply leads to 1 additional element in `data`, `indices` and `indptr` of the csr representation, which seems rather minimal to me. And I see an advantage in using mainline csr (i'd be tempted to use csc) representations as we do not need to implement and optimize matrix operations on our own.

One other thing regarding this class and its `T_small` attribute is this:

>This package [`scipy.sparse`] is switching to an array interface, compatible with NumPy arrays, from the older matrix interface. We recommend that you use the array objects ([bsr_array](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.bsr_array.html#scipy.sparse.bsr_array), [coo_array](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy.sparse.coo_array), etc.) for all new work.

_[source](https://docs.scipy.org/doc/scipy/reference/sparse.html)_

I'm unfamiliar with what exactly this compatibility with numpy arrays entails. But we could, while we're at it, move away from `csr_matrix`.

**Comment by alexbovet** [comment]
The reason I created this class is that I observed that when you have very large stochastic matrices (~100'000 nodes), which are actually very sparse the scipy implementation of csr-csr matmul was pretty slow. I realize that since the matrices are very sparse, there are actually just a few multiplications that arise from the off-diagonal elements, but because of the non-zero diagonal, there are still many multiplications to do (most of them being 1 x 1)... So you can decompose the problem to avoid all those multiplications.

This is explained in the __matmul__ I wrote:

# We split the problem in A @ B = (A-aI) @ (B-bI) + b*A + a*B - a*b*I
# where a is the diag_val of A and b is the diag_val of B
# Moreover, if intersection(A.nz_rowvals,B.nz_rowvals) = [],
# (A-aI) @ (B-bI) = 0 and there is no multiplication to do

Since I only use this to multiply inter-transition matrices, there is no need for __rmatmul__ or the other operations.

Now, I have to say that I did not test this systematically and it may be that with newer scipy versions, this is not a problem anymore. Also, my implementation may introduce too much overhead compared to a simple csr matmul. So it may very well be that we could get rid of this. 

This would need some testing on large stochastic matrices. But anyway, the code also works if one uses scipy sparse matrices instead. So this is not so urgent.

Regarding the compatibility with the future scipy sparse array, this is a good thing because they were still using numpy matrices which are a pain to use. I use csr_matrix in many places, we can simply move to csr_array. The change should not be too complicated (I already use @ for the matrix multiplication).  I would keep csr instead of csc, because we have a lot of row-stochastic matrices and we often need to normalize the row sum to one).

**Comment by j-i-l** [comment]
I see. This makes sense to me. Thanks for the explanation.

I'll setups some simple tests to get a better idea of the performance for `__matmul__` (will do this in #35).

The switch to `csr_array` we best address in a separate issue.

**Comment by alexbovet** [comment]
Sorry, I forgot to mention another place where this class is useful: when [computing the inter-transition matrices](https://github.com/alexbovet/flow_stability/blob/8648a8bd715e3b086ed13cb7453d41754c176398/src/flowstab/TemporalNetwork.py#L1288C1-L1288C87), if you set `use_sparse_stoch=True` it will decompose the adjacency matrix to find the connected component and compute the transition matrix on each component separately with [sparse_lapl_exp](https://github.com/alexbovet/flow_stability/blob/8648a8bd715e3b086ed13cb7453d41754c176398/src/flowstab/TemporalNetwork.py#L2473). Since laplacians are usually sparse, taking the matrix exponential on smaller components separately is usually much faster than doing it on the whole thing. There is even the option to do each component in parallel actually.

If you look at the example of the [mice network](https://github.com/alexbovet/flow_stability/blob/main/examples/wild_mice_network.ipynb), it is indeed much faster, even for small networks.
